phylum/connector.lisp

```diff
@@ -0,0 +1,215 @@
+;; context, along with the OID and class factory. The class factory is used
+;; to instiante the object for that OID, and call the handle method on that
+;; object with the response.
+;;
+;; Upon execution of the callback, the original event context is purged, along
+;; with the original request (or deleted if in state DB), so as to reduce
+;; storage space.
+;;
+(defun mk-connector-handler ()
+  (let ([state (sorted-map)])
+    (labels
+      ([register-handler (name fn)
+         (assoc! state name fn)]
+
+       [mk-request-key (req-id)
+         (format-string "$cr:{}" req-id)]
+
+       ;; IMPORTANT: we store this in the central `private` PDC since we
+       ;; don't know which PDC the original request was stored.
+       [register-request-callback (req-id handler-name &optional ctx)
+         (let* ([ctx (default ctx (sorted-map))]
+                [callback-state (sorted-map
+                               "handler_name" handler-name)])
+           (assoc! ctx "request_id" req-id)
+           (assoc! callback-state "ctx" ctx)
+           (sidedb:put (mk-request-key req-id) callback-state))]
+
+       [unregister-request (req-id ctx)
+         ;; TODO: handle multiple responses for single request_id (optional) 
+         (let* ([callback-state-key (mk-request-key req-id)]
+                [key (get ctx "key")]
+                [pdc (get ctx "pdc")])
+           (sidedb:purge callback-state-key)
+           (if pdc
+            (cc:storage-purge-private pdc key)
+            (statedb:del key)))]
+
+       [get-callback-state (req-id)
+         (sidedb:get (or (mk-request-key req-id)
+                         (error 'missing-handler "no registered handler")))]
+
+       [call-handler-helper (resp-body)
+         (unless resp-body (error 'missing-resp "missing response"))
+         (let* ([req-id (get resp-body "request_id")]
+                [callback-state (get-callback-state req-id)]
+                [handler-name (get callback-state "handler_name")]
+                [ctx (default (get callback-state "ctx") (sorted-map))]
+                [msp (get ctx "msp")]
+                [handler-fn (get state handler-name)])
+           (when msp 
+             (cc:debugf (sorted-map "msp" msp "req_id" req-id) "validating MSP")
+             (if (valid-msp? msp)
+               (cc:debugf (sorted-map "msp" msp) "MSPID validated")
+               (set-exception-security "invalid MSP for response")))
+           (if handler-fn
+             (handler-fn resp-body ctx)
+             (error 'missing-handler 
+                    (format "missing connector handler: {}" handler-name)))
+           (unregister-request req-id ctx))]
+
+          [call-handler-helper-recurse-i (i)
+            (let* ([resp-body (transient:get (format-string "$ch_rep:{}" i))])
+              (when resp-body
+                (call-handler-helper resp-body)
+                (call-handler-helper-recurse-i (+ i 1))))]
```

Comment:
Indentation seems off here.

Wouldn't this be pretty inefficient? Does the index ever reset?

Reply:
will fix indentation. 

Don't quite follow the question: the index pretty quickly reaches `nil` and aborts. The recurse call is the last line of the function so it will turn into a forloop (tail call optimization) underneath the hood.

Reply:
I didn't really follow what was going on with the index or how it is bounded.  It seems to be stored under a global key which seemed odd.

Reply:
There's no index (to save space). It merely checks a well known list of keys, starting with `$ch_rep:0` ... `$ch_repo:N` until it gets a nil value, then aborts. In most cases it stops after `$ch_rep:1`, since there's only a single response at `$ch_repo:0`.

Reply:
Ah I see what was confusing me.  I missed that this was transient data.  Also read that when statement as an if basically which was quite confusing.  Seems okay then.